lewis/adapters/stream.py

```diff
@@ -163,8 +186,12 @@ def __init__(self, func, pattern, argument_mappings=None, return_mapping=None, d
             raise RuntimeError('Can not construct a Func-object from a non callable object.')
 
         self.func = func
-        self.raw_pattern = pattern
-        self.pattern = re.compile(b(pattern), 0) if pattern else None
+
+        self.raw_pattern = pattern.raw_pattern
+        self.pattern = re.compile(b(pattern.regex_pattern), 0)
```

Comment:
Regarding, 3. I also noticed that during experimentation...I'm not sure if that makes it better or not, because I'm not sure any more why we have `^` in all *our* regexes (probably because "we've always done it like this"). And because we would basically force people to add an extra character that is not required (although to be honest I assumed it was), but in the case of scanf, where a literal `^` at the beginning is as valid as any other character it would require `fmt(...)`. Kind of also touches point 2 from my list above.

Concerning 2, I agree with you in terms of "should be a logic error"...and from number 3 follows, that this has to be done already now.

Number 4 is totally unexpected and I never saw anything of that magnitude coming :)

I think there are very good reasons to go either way and I'm not totally convinced about either yet, so let's think about a way to make this better.

In the direction of auto-detection, are there other options than using `^`? Is it enough to look for `%([fdix]{1})` to guess with a high enough probability that it's a scanf pattern and treat it as a regex otherwise? For the patterns without such identifiers it would not be relevant except for escaping reasons. That still wouldn't solve the "extensibility" issue though.

In the other direction, could there be an easier way to define a sort of default pattern type? Could it be made "more automatic"?